### PR TITLE
BootloaderCore: Handle BIST failure in SecEntry

### PR DESCRIPTION
**Pull request by CunningLearner**
- add BistVal in STAGE1A_ASM_HOB structure
- use MM0 register to preserve the BIST value
- push BistVal while setup HOB in stack
- add check for CPU BIST failure and halt the system when failed

Signed-off-by: Himanshu Sahdev aka CunningLearner <sahdev.himan@gmail.com>

### CODE REVIEW

File: BootloaderCorePkg/Stage1A/Ia32/SecEntry.nasm [hunk 1]
```diff
@@ -62,6 +64,7 @@ FspApiSuccess:
         mov     eax, esp
 
         ; Setup HOB
+        push    ebx                  ; BistVal
```
Comment (myguitar) [inline BootloaderCorePkg/Stage1A/Ia32/SecEntry.nasm]: I don't think EBX is preserved here. Only EDI and ESI are preserved. You may need to use EDI. Can you also add a debug message in Stage1A.c to see if BistVal is expected one?

Comment (mauricema) [reply BootloaderCorePkg/Stage1A/Ia32/SecEntry.nasm]: I think we can put it back into MM0 again.  
FSP 2.0 TempRamInit() API preserves the following general purpose registers EBX, EDI, ESI, EBP and the following floating point registers MM0, MM1.

Comment (mauricema) [reply BootloaderCorePkg/Stage1A/Ia32/SecEntry.nasm]: Agree.  if we see a CPU BIST error,  we might want to print this error and then Halt the CPU.  This needs to be done after the serial port initialization.

Comment (CunningLearner) [reply BootloaderCorePkg/Stage1A/Ia32/SecEntry.nasm]: Thanks for the valuable points, I used MM0 register to hold the value and as it did not need to be accessed by pushing and popping in the same way as the FPU registers, I pushed the EBX register. Also, I debug the same as @myguitar suggested and confirmed with the expected value, and put a valid check and halt the CPU when BIST failure!

File: BootloaderCorePkg/Stage1A/Stage1A.c [hunk 2]
```diff
@@ -240,6 +240,9 @@ SecStartup2 (
     DEBUG ((DEBUG_INIT, "\n%a\n", mBootloaderName));
   }
 
+  if(Stage1aAsmHob->BistVal != 0)
```
Comment (myguitar) [inline BootloaderCorePkg/Stage1A/Stage1A.c]: a space between if and (.

if (Stage1aAsmHob->BistVal != 0) {
  CpuHalt ("BIST failure!\n");
}

File: BootloaderCorePkg/Stage1A/Stage1A.c [hunk 3]
```diff
@@ -240,6 +240,9 @@ SecStartup2 (
...
+    CpuHalt ("BIST failure!\n");
```
Comment (mauricema) [inline BootloaderCorePkg/Stage1A/Stage1A.c]: We can make it more specific by saying "CPU BIST failure!"

File: BootloaderCorePkg/Stage1A/Ia32/SecEntry.nasm (hunk 1 above)
Comment (mauricema) [reply BootloaderCorePkg/Stage1A/Ia32/SecEntry.nasm]: Line 64 "mov     eax, esp", the "eax" is not used at all after the assignment.   So let us avoid another register "ebx" just for pushing it into stack.   I think we can safely remove line 64 and use "eax" to push into stack.

File: BootloaderCorePkg/Stage1A/Stage1A.c (hunk 3 above)
Comment (CunningLearner) [reply BootloaderCorePkg/Stage1A/Stage1A.c]: Ack

File: BootloaderCorePkg/Stage1A/Stage1A.c (hunk 2 above)
Comment (CunningLearner) [reply BootloaderCorePkg/Stage1A/Stage1A.c]: Done

File: BootloaderCorePkg/Stage1A/Ia32/SecEntry.nasm [hunk 4]
```diff
@@ -61,7 +63,11 @@ FspApiSuccess:
         add     esp, dword [ASM_PFX(PcdGet32(PcdStage1StackSize))]
         mov     eax, esp
 
+        movd    ebx, mm0
```
Comment (mauricema) [inline BootloaderCorePkg/Stage1A/Ia32/SecEntry.nasm]: I did put a comment in previous commit.  Can we remove line 64  and then do  below ?
movd  eax, mm0
push   eax

File: BootloaderCorePkg/Stage1A/Ia32/SecEntry.nasm (hunk 1 above)
Comment (CunningLearner) [reply BootloaderCorePkg/Stage1A/Ia32/SecEntry.nasm]: It was actually hit me yesterday, but I was not sure why it was actually there, maybe intentionally!?

Well, I successfully boot with eax in place, implies we can remove it!

File: BootloaderCorePkg/Stage1A/Ia32/SecEntry.nasm (hunk 4 above)
Comment (CunningLearner) [reply BootloaderCorePkg/Stage1A/Ia32/SecEntry.nasm]: Done!

File: BootloaderCorePkg/Stage1A/Stage1A.c [hunk 5]
```diff
@@ -240,6 +240,9 @@ SecStartup2 (
     DEBUG ((DEBUG_INIT, "\n%a\n", mBootloaderName));
   }
 
+  if (Stage1aAsmHob->BistVal != 0)
+    CpuHalt ("CPU BIST failure!\n");
```
Comment (mauricema) [inline BootloaderCorePkg/Stage1A/Stage1A.c]: Minor comments on coding styles.  :)
According to EDK2 coding style, it is required to add curly brace {} after if condition even if it has only single statement after if.  So it needs to be:
if (Stage1aAsmHob->BistVal != 0) {
    CpuHalt ("CPU BIST failure!\n");
}
Aiden also pointed it out in his previous comments.

Comment (CunningLearner) [reply BootloaderCorePkg/Stage1A/Stage1A.c]: Done

**Pull request merged**